Infinite 'Advancing pipeline' loop: writeBlockerPlaceholder skips DB update, re-dispatch inherits stale recoveryAttempts

## Problem

When an \`execute-task\` unit times out and recovery exhausts all retries, auto-mode writes a blocker placeholder and logs \"Advancing pipeline\" — but then logs the same message 8–9 more times in a tight loop without ever actually advancing. The loop only terminates when Doctor clears the stale lock.

Observed output:
```
Warning: execute-task M007/S02/T02 skipped after 2 recovery attempts (task checkbox unchecked in
.gsd/milestones/M007/slices/S02/S02-PLAN.md; state next action still points at the timed-out task).
Blocker artifacts written. Advancing pipeline. (attempt 1)
...
Warning: execute-task M007/S02/T02 skipped after 2 recovery attempts ... Advancing pipeline. (attempt 9)
```

## Root Cause

Two bugs compound to produce the infinite loop.

---

### Bug 1 (new — this issue): `writeBlockerPlaceholder` does not update the DB task status

**File:** `auto-recovery.js`, function `writeBlockerPlaceholder` (~line 398)

For `execute-task`, the placeholder writes `T{tid}-SUMMARY.md` to disk but never calls `updateTaskStatus()`. However, `verifyExpectedArtifact` for `execute-task` uses the DB as the authoritative source when available:

```js
// auto-recovery.js ~line 274
const dbTask = getTask(mid, sid, tid);
if (dbTask) {
    if (dbTask.status !== "complete" && dbTask.status !== "done")
        return false;  // DB says "pending" → verification fails
}
```

Because the DB task status stays `"pending"`, `verifyExpectedArtifact` returns `false`. The task is never added to `s.completedUnits`, `deriveState` re-derives the same task as active (checkbox still unchecked, DB still pending), dispatch returns the same unit, and the loop repeats.

**Fix:**
```js
// auto-recovery.js — writeBlockerPlaceholder()
// After writing the file, mark the DB task complete so verifyExpectedArtifact passes:
if (unitType === "execute-task" && isDbAvailable()) {
    const [mid, sid, tid] = unitId.split("/");
    if (mid && sid && tid) {
        try { updateTaskStatus(mid, sid, tid, "complete", Date.now()); }
        catch { /* non-fatal */ }
    }
}
```

---

### Bug 2 (amplifier — also tracked in closed/deferred #2322): `recoveryAttempts` not reset on re-dispatch

**File:** `auto/phases.js`, function `runUnitPhase` (~line 594)

The dispatch-time runtime record write does not include `recoveryAttempts: 0`:

```js
deps.writeUnitRuntimeRecord(s.basePath, unitType, unitId, s.currentUnit.startedAt, {
    phase: "dispatched",
    wrapupWarningSent: false,
    timeoutAt: null,
    lastProgressAt: s.currentUnit.startedAt,
    progressCount: 0,
    lastProgressKind: "dispatch",
    // recoveryAttempts: MISSING — inherited from prior dispatch via merge logic
});
```

**File:** `unit-runtime.js` ~line 34 (merge logic):
```js
recoveryAttempts: updates.recoveryAttempts ?? prev?.recoveryAttempts ?? 0,
```

After the first skip writes `recoveryAttempts: 2` (idle, maxRecoveryAttempts=2), subsequent dispatches inherit the value. The next idle timeout fires, sees `2 >= 2`, and immediately skips again — no steering, no wait. `unitRecoveryCount` is deleted on each skip and resets to 0, so the 30s backoff in `recoverTimedOutUnit` never fires.

**Fix:**
```js
deps.writeUnitRuntimeRecord(s.basePath, unitType, unitId, s.currentUnit.startedAt, {
    phase: "dispatched",
    wrapupWarningSent: false,
    timeoutAt: null,
    lastProgressAt: s.currentUnit.startedAt,
    progressCount: 0,
    lastProgressKind: "dispatch",
    recoveryAttempts: 0,       // ← ADD
    lastRecoveryReason: null,  // ← ADD
});
```

---

## Why Both Must Be Fixed Together

Even if Bug 2 is fixed (each re-dispatch gets a fresh recovery budget), Bug 1 alone causes the task to loop indefinitely — the dispatch rules will keep returning the same task because the DB status and plan checkbox are never updated. Bug 2 just makes the loop faster and swallows the steering messages.

## Expected Behavior

After `writeBlockerPlaceholder` runs for an `execute-task`:
1. DB task status is set to `"complete"`
2. `verifyExpectedArtifact` returns `true`
3. Task is added to `s.completedUnits`
4. `deriveState` advances to the next task or phase
5. "Advancing pipeline" appears exactly once

## Environment

- GSD version: 2.44.0
- Model: claude-sonnet-4-6
- Unit: execute-task/M007/S02/T02
- Dispatched: 4 times (forensic anomaly: stuck-loop ERROR)
- "Advancing pipeline" logged: 9 times

## Reproduction Context

MCP tool call (`foldphone.create_device`) timed out with `-32001: Request timed out` during task execution. The unit exceeded the idle timeout, recovery was attempted twice (idle mode, maxRecoveryAttempts=2), a blocker placeholder was written, and the pipeline entered the infinite advance loop. Doctor cleared the stale lock but the underlying state was not repaired.

## Forensic Evidence

- `[ERROR] stuck-loop`: Unit execute-task/M007/S02/T02 was dispatched 4 times
- `[WARNING] error-trace`: MCP call failed: foldphone.create_device — MCP error -32001: Request timed out (×4 in activity logs)
- 9× "Advancing pipeline" messages in TUI output
- Crash lock left active with PID alive
- Task checkbox unchecked in S02-PLAN.md (confirmed by recovery status message)
- No SUMMARY written by the agent (blocker placeholder written by recovery instead)

## Recovery (Immediate)

The DB task is stuck at `"pending"` with a blocker placeholder SUMMARY on disk. Fix with:

```bash
sqlite3 .gsd/gsd.db "UPDATE tasks SET status='complete', completed_at=$(date +%s)000 WHERE milestone_id='M007' AND slice_id='S02' AND id='T02';"
```

Then mark the checkbox in `.gsd/milestones/M007/slices/S02/S02-PLAN.md` and run `/gsd auto` to resume.

## Related

- #2322 (closed/deferred): `recoveryAttempts` persists across re-dispatches — this issue is the prerequisite for Bug 2 above, deferred pending arch rewrite in #2041.

---
*Auto-generated by `/gsd forensics`*

--- a/src/resources/extensions/gsd/auto-recovery.ts
+++ b/src/resources/extensions/gsd/auto-recovery.ts
@@ -12,7 +12,7 @@ import { parseUnitId } from "./unit-id.js";
 import { atomicWriteSync } from "./atomic-write.js";
 import { clearParseCache } from "./files.js";
 import { parseRoadmap as parseLegacyRoadmap, parsePlan as parseLegacyPlan } from "./parsers-legacy.js";
-import { isDbAvailable, getTask, getSlice, getSliceTasks } from "./gsd-db.js";
+import { isDbAvailable, getTask, getSlice, getSliceTasks, updateTaskStatus } from "./gsd-db.js";
 import { isValidationTerminal } from "./state.js";
 import {
   nativeConflictFiles,
@@ -425,6 +425,20 @@ export function writeBlockerPlaceholder(
     `Review and replace this file before relying on downstream artifacts.`,
   ].join("\n");
   writeFileSync(absPath, content, "utf-8");
+
+  // Mark the task as complete in the DB so verifyExpectedArtifact passes.
+  // Without this, the DB status stays "pending" and the dispatch loop
+  // re-derives the same task indefinitely (#2531).
+  if (unitType === "execute-task" && isDbAvailable()) {
+    const parts = unitId.split("/");
+    const mid = parts[0];
+    const sid = parts[1];
+    const tid = parts[2];
+    if (mid && sid && tid) {
+      try { updateTaskStatus(mid, sid, tid, "complete", new Date().toISOString()); } catch { /* non-fatal */ }
+    }
+  }
+
   return diagnoseExpectedArtifact(unitType, unitId, base);
 }
 

--- a/src/resources/extensions/gsd/tests/idle-recovery.test.ts
+++ b/src/resources/extensions/gsd/tests/idle-recovery.test.ts
@@ -294,3 +294,66 @@ test('verifyExpectedArtifact: hook types always return true', () => {
   }
 });
 
+
+test('writeBlockerPlaceholder: updates DB task status for execute-task (#2531)', async () => {
+  const base = createFixtureBase();
+  try {
+    const { openDatabase, closeDatabase, insertMilestone, insertSlice, insertTask, getTask, isDbAvailable } =
+      await import("../gsd-db.ts");
+
+    const dbPath = join(base, ".gsd", "gsd.db");
+    // Create the tasks directory (required for artifact path resolution)
+    mkdirSync(join(base, ".gsd", "milestones", "M001", "slices", "S01", "tasks"), { recursive: true });
+
+    openDatabase(dbPath);
+    try {
+      insertMilestone({ id: "M001", title: "Test", status: "active" });
+      insertSlice({ id: "S01", milestoneId: "M001", title: "Slice", status: "active" });
+      insertTask({ id: "T01", sliceId: "S01", milestoneId: "M001", title: "Task", status: "pending" });
+
+      // Before fix: writeBlockerPlaceholder wrote the file but left DB as "pending"
+      writeBlockerPlaceholder("execute-task", "M001/S01/T01", base, "idle recovery exhausted");
+
+      const task = getTask("M001", "S01", "T01");
+      assert.equal(task?.status, "complete",
+        "writeBlockerPlaceholder must update DB task status to 'complete' so verifyExpectedArtifact passes");
+
+      // Verify the full chain works: verifyExpectedArtifact should return true
+      const verified = verifyExpectedArtifact("execute-task", "M001/S01/T01", base);
+      assert.equal(verified, true,
+        "verifyExpectedArtifact should pass after writeBlockerPlaceholder updates DB status");
+    } finally {
+      if (isDbAvailable()) closeDatabase();
+    }
+  } finally {
+    cleanup(base);
+  }
+});
+
+test('writeBlockerPlaceholder: does NOT update DB for non-execute-task types', async () => {
+  const base = createFixtureBase();
+  try {
+    const { openDatabase, closeDatabase, insertMilestone, insertSlice, getSlice, isDbAvailable } =
+      await import("../gsd-db.ts");
+
+    const dbPath = join(base, ".gsd", "gsd.db");
+    mkdirSync(join(base, ".gsd", "milestones", "M001", "slices", "S01"), { recursive: true });
+
+    openDatabase(dbPath);
+    try {
+      insertMilestone({ id: "M001", title: "Test", status: "active" });
+      insertSlice({ id: "S01", milestoneId: "M001", title: "Slice", status: "active" });
+
+      // research-slice is NOT execute-task — DB should NOT be updated
+      writeBlockerPlaceholder("research-slice", "M001/S01", base, "idle recovery exhausted");
+
+      const slice = getSlice("M001", "S01");
+      assert.equal(slice?.status, "active",
+        "writeBlockerPlaceholder should not change DB status for non-execute-task types");
+    } finally {
+      if (isDbAvailable()) closeDatabase();
+    }
+  } finally {
+    cleanup(base);
+  }
+});

